File: examples/looker_embed_generator/looker_embed_generator.java
/**
 *Class          : looker_embed_generator.java
 *
 *Description    : Sample Java class to generate secure Looker
 *               : embed codes that include custom filters.
 * 
 *Requirements   : You must set the following environment variable
 *               : to establish credentials for the embed user:
 * 
 *               : export LOOKERKEY=<<Looker embed key>>   ## bash
 *               : set LOOKERKEY=<<Looker embed key>>      :: cmd
 *               : $env:LOOKERKEY = "<<Looker embed key>>" ## powershell 
 * 
 *Usage          : java looker_embed_generator <<environment>> -e <<embed_url>> -u [<<attribute>> <<filter>>]
 *               :
 *               : eg: java looker_embed_generator.java prod dashboards/18
 *               :     java looker_embed_generator.java prod looks/98 browser Chrome
 *               : To create an embed string with filter(s):
 *               :
 *               : 1 Filter and 1 Value
 *               :
 *               : java looker_embed_generator.java prod dashboards/18
 *               :   '{"filter-name": "filtername-value",  "matchtype": "matchtype-value", 
 *               :     "matchtype": "matchtype-value", "values":"filter-value"}'
 *               :
 *               : eg: java looker_embed_generator.java prod dashboards/18
 *               :   '{"filterName":"City","matchType":"=","matchValue":"Metropolis"}'
 *               :
 *               :
 *               : NOTE: The embed must be accessible to the 
 *               : Embed Shared Group.
 *               :
 *References     : https://github.com/looker/looker_embed_sso_examples

 *               : https://docs.looker.com/reference/embedding/sso-embed
 *               : https://docs.looker.com/reference/embedding/embed-javascript-events
 * 
*/

import java.math.BigInteger;
import java.security.SecureRandom;
import java.util.Calendar;
import java.util.Date;
import java.util.Map;
import java.util.Base64;
import java.util.List;
import java.util.Arrays;
import java.io.*;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import java.net.URLEncoder;

import java.nio.charset.StandardCharsets;


public class looker_embed_generator {

    public static void main(String [] args){

        // Get CMD Line Arguments as a List
        List<String> arguments = Arrays.asList(args);

        // dev.analytics.gov
        String lookerURL = "";  // to be assigned by <<environment>> argument
        String lookerKey = ""; // will be set by the system environment variable LOOKERKEY
        String externalUserID = "50";  // converted to JSON string
        String firstName = "\"Dashboard\""; // converted to JSON string
        String lastName = "\"User\""; // converted to JSON string
        String permissions = "[\"see_lookml_dashboards\", \"access_data\", \"see_user_dashboards\", \"see_looks\"]"; // converted to JSON array
        String models = "[\"all\"]"; // converted to JSON array
        String groupIDs = "[]"; // converted to JSON array, can be set to null (value, not JSON) for no groups
        String externalGroupID = "\"external_group_id\"";  // converted to JSON string
        String sessionLength = "900";
        String embedURL = "";
        String forceLogoutLogin = "true"; // converted to JSON bool
        String accessFilters = ("{}");  // converted to JSON Object of Objects
        String userAttributes = "{\"can_see_sensitive_data\": \"YES\"}";  // A Map<String, String> converted to JSON object
        String queryString = "";

        lookerKey = System.getenv("LOOKERKEY");
        if (lookerKey == null) {
            System.err.println("Environment variable LOOKERKEY doesn't exist");
            System.exit(1);
        }

        if (args.length < 2) {
            System.err.println("Usage: java looker_embed_generator <<environment>> <<embed_url>> [<<attribute>> <<filter>>]");
            System.exit(1);
        }

        if (args[0].equalsIgnoreCase("PROD") || args[0].equalsIgnoreCase("PRODUCTION")) {
            lookerURL = "analytics.gov.bc.ca";
        } else if (args[0].equalsIgnoreCase("DEV") || args[0].equalsIgnoreCase("DEVELOPMENT")){
            lookerURL = "dev.analytics.gov.bc.ca";
        } else {
            System.err.println("Usage: java looker_embed_generator <<environment>> <<embed_url>> [<<attribute>> <<filter>>]");
            System.exit(1);
        }

        // build query string from cmd line parameters
        if (args.length > 2 && arguments.contains("-e")) {
            // Check that the argument for the embed filter is non-zero length
            if (args[arguments.indexOf("-e")+1].length() != 0) {
                try {
                    queryString = createQueryString(args[3]);
                    embedURL = "/embed/" + args[1] + queryString + "&embed_domain=http://127.0.0.1:5000";
                } catch (Exception e) {
                    System.out.println(e);
                }
            }
        } else {
            // No query string to add, generate embedURL without it.
            embedURL = "/embed/" + args[1] + "?embed_domain=http://127.0.0.1:5000";
        }

        // Add the user attribute if that parameter is set in the arguments
        if (args.length > 2 && arguments.contains("-u")) {
            // Check that the argument for the user attribute is non-zero length
            if (args[arguments.indexOf("-u")+1].length() != 0 ) { 
                try {
                    // adding a new attribute and filter value to the userAttributes JSON blob
                    String attribute = ", \"" + args[arguments.indexOf("-u")+1] + "\": \"" + args[arguments.indexOf("-u")+2] + "\"";
                    StringBuilder newUserAttributes = new StringBuilder(userAttributes);
                    userAttributes = newUserAttributes.insert(userAttributes.length()-1, attribute).toString();
                } catch (Exception e) {
                    System.out.println(e);
                }
            }
        }

        try {

            String url = createURL(lookerURL, lookerKey, externalUserID, firstName, lastName, permissions, models,
                                   sessionLength, accessFilters, embedURL, forceLogoutLogin, groupIDs,
                                   externalGroupID, userAttributes);
            System.out.println("https://" + url);

        } catch(Exception e){
            System.out.println(e);
        }
    }

    // Takes filter parameters and creates a string to pass in with the embed string
    public static String createQueryString(String params) {

        Filter f = new Filter();
        Gson gson = new Gson();
        f = gson.fromJson(params, f.getClass());
        
        String filterName = f.filterName;
        String matchType = f.matchType;
        String matchValue = f.matchValue;
        String encodedFilterString = "";

        try {
            // dont actually know if i need this
            encodedFilterString =  URLEncoder.encode(f.matchValue, StandardCharsets.UTF_8);
            encodedFilterString = encodedFilterString.replace("+", "%20");
        } catch(Exception e){
            System.out.println(e);
        }

        String queryString =
            "?filter_config={\"" + filterName +
            "\":[{\"type\":\"" + f.matchType +  "\",\"values\":[{\"constant\":\"" + f.matchValue +
            "\"},{}],\"id\":456}]}";

        return queryString;
    }

    // builds the embed URL from the parameter
    public static String createURL(String lookerURL, 
                                   String lookerKey,
                                   String userID, 
                                   String firstName, 
                                   String lastName, 
                                   String userPermissions,
                                   String userModels, 
                                   String sessionLength, 
                                   String accessFilters,
                                   String embedURL, 
                                   String forceLogoutLogin, 
                                   String groupIDs,
                                   String externalGroupID, 
                                   String userAttributes) 
    throws Exception {

        String path = "/login/embed/" + java.net.URLEncoder.encode(embedURL, "UTF-8");

        Calendar cal = Calendar.getInstance();
        SecureRandom random = new SecureRandom();
        String nonce = "\"" + (new BigInteger(130, random).toString(32)) + "\"";  // converted to JSON string
        String time = Long.toString(cal.getTimeInMillis() / 1000L);

        // Order of these here is very important! See:
        // https://docs.looker.com/reference/embedding/sso-embed#signature
        String urlToSign = "";
        urlToSign += lookerURL + "\n";
        urlToSign += path + "\n";
        urlToSign += nonce + "\n";
        urlToSign += time + "\n";
        urlToSign += sessionLength + "\n";
        urlToSign += userID + "\n";
        urlToSign += userPermissions + "\n";
        urlToSign += userModels + "\n";
        urlToSign += groupIDs + "\n";
        urlToSign += externalGroupID + "\n";
        urlToSign += userAttributes + "\n";
        urlToSign += accessFilters;

        String signature =  encodeString(urlToSign, lookerKey);

        // you need to %20-encode each parameter before you add to the URL
        String signedURL = "nonce="    + java.net.URLEncoder.encode(nonce, "UTF-8") +
                "&time="               + java.net.URLEncoder.encode(time, "UTF-8") +
                "&session_length="     + java.net.URLEncoder.encode(sessionLength, "UTF-8") +
                "&external_user_id="   + java.net.URLEncoder.encode(userID, "UTF-8") +
                "&permissions="        + java.net.URLEncoder.encode(userPermissions, "UTF-8") +
                "&models="             + java.net.URLEncoder.encode(userModels, "UTF-8") +
                "&access_filters="     + java.net.URLEncoder.encode(accessFilters, "UTF-8") +
                "&signature="          + java.net.URLEncoder.encode(signature, "UTF-8") +
                "&first_name="         + java.net.URLEncoder.encode(firstName, "UTF-8") +
                "&last_name="          + java.net.URLEncoder.encode(lastName, "UTF-8") +
                "&group_ids="          + java.net.URLEncoder.encode(groupIDs, "UTF-8") +
                "&external_group_id="  + java.net.URLEncoder.encode(externalGroupID, "UTF-8") +
                "&user_attributes="    + java.net.URLEncoder.encode(userAttributes, "UTF-8") +
                "&force_logout_login=" + java.net.URLEncoder.encode(forceLogoutLogin, "UTF-8");

        return lookerURL + path + '?' + signedURL;

    }

    // return a MAC of the input string cryptographically signed with the LOOKERKEY secret
    public static String encodeString(String stringToEncode, String lookerKey) throws Exception {
        byte[] keyBytes = lookerKey.getBytes();
        SecretKeySpec signingKey = new SecretKeySpec(keyBytes, "HmacSHA1");
        Mac mac = Mac.getInstance("HmacSHA1");
        mac.init(signingKey);
        byte[] rawHmac = Base64.getEncoder().encode(mac.doFinal(stringToEncode.getBytes("UTF-8")));
        return new String(rawHmac, "UTF-8");
    }
}

    public class Filter {

        public static void main (String [] args) {}

        public String matchType = "";
        public String filterName = "";
        public String matchValue = "";
    } 

File: examples/looker_embed_generator/looker_embed_generator.py
###################################################################
# Script Name   : looker_embed_generator.py
#
# Description   : Sample Python script to generate secure Looker
#               : embed codes that include custom filters.
#
# Requirements  : You must set the following environment variable
#               : to establish credentials for the embed user
#
#               : export LOOKERKEY=<<Looker embed key>>
#
# Usage         : To create an embed string without a filter:
#               : python looker_embed_generator.py <<embed url>>
#               :
#               : eg: python looker_embed_generator.py dashboards/18
#               :
#               : To create an embed string with filter(s):
#               :
#               : 1 Filter and 1 Value
#               :
#               : python looker_embed_generator.py <<embed url>>
#               :   '{"filter-name": {"matchtype": "matchtype-value", 
#               :                     "values":"filter-value"}'
#               :
#               : eg: python looker_embed_generator.py dashboards/18
#               :   '{"City":["matchtype":"=", "values":"Metropolis"]}'
#               :
#               : or:
#               : 
#               : 1 Filter and 2 or more Values
#               :
#               : python looker_embed_generator.py <<embed url>>
#               :   '{"filter-name": ["matchtype": {"matchtype-value","values":"filter-value"},
#               :                                  {"matchtype-value","values":"filter-value"}]}'
#               :
#               : eg: python looker_embed_generator.py dashboards/18
#               :   '{"City":[{"matchtype":"=", "values":"Duncan - British Columbia"},
#               :             {"matchtype":"starts", "values":"Nanaimo - British Columbia"}]}'
#               :
#               : or:
#               : 
#               : 2 Filter and 1 Value
#               :
#               : python looker_embed_generator.py <<embed url>>
#               :   '{"filter-name": ["matchtype": {"matchtype-value","values":"filter-value"}],
#               :    "filter-name": ["matchtype": {"matchtype-value","values":"filter-value"}]}'
#               :
#               : eg: python looker_embed_generator.py dashboards/snowplow_web_block::bc_gov_analytics_dashboard_1 
#               :   '{"City":[{"matchtype":"=","values":"Victoria - British Columbia"}, 
#               :     "City":[{"matchtype":"contains","values":"Vancouver - British Columbia"}]}'
#               :
#               : or:
#               : 
#               : 1 Filter and comma separated values
#               :
#               : python looker_embed_generator.py <<embed url>>
#               :   '{"filter-name": ["matchtype": "matchtype-value", 
#               :                     "values":"filter-value"]}'
#               :
#               : :python looker_embed_generator.py dashboards/18
#               :   '{"Theme": ["matchtype": "=", 
#               :               "values":"\"Birth, Adoption, Death, Marriage & Divorce\",\"Employment, Business & Economic Development\""]}'
#               :
#               :
#               : List of valid match types:
#               :
#               : Is equal to: '='
#               : Contains: 'contains'
#               : Does not contain: '!contains'
#               : Starts with: 'starts'
#               : Is empty: 'empty'
#               : Is null: 'null'
#               : Is not equal to: '!='
#               : Does not contain: '!contains'
#               : Does not start with: '!starts'
#               : Does not end with: '!ends'
#               : Is not empty: '!empty'
#               : Is not null: '!null'
#               : Matches a user attribute: 'user_attribute'
#               :
#               :
#               : NOTE: The embed must be accessible to the
#               : Embed Shared Group.
#               :
# References    : https://docs.looker.com/reference/embedding/
#               :   sso-embed#building_the_embed_url
#               : https://docs.looker.com/reference/embedding/
#               :   embed-javascript-events
#               :

import urllib
import base64
import json
import time
import binascii
import os
from hashlib import sha1
import hmac
import sys  # to read command line parameters


# Add double slashes before commas and then URI encode.
# The double slashes are required due to how looker parses
# filter parameters in the embed_url string.
def parse_filter_value(filter_value):
    parsed_filter_value = urllib.quote(filter_value.replace(', ', r'\\, ').replace('\"',''))
    return parsed_filter_value


# This function loops over the filters passed into the script and
# generates the filters section of the embed url string.
def encode_embed_filters(filters):
    embed_url_string = "?filter_config=%7B"
    filter_string = ''
    index = 1
    for filter_name in filters:
        filter_string += "\"" + filter_name + "\":%5B"
        filter_string += build_filter_string(filters[filter_name])
        if index < len(filters):
            filter_string += ","
        index += 1
    return embed_url_string + filter_string + "%7D"


# This function loops over each filter value and generates the
# the strings for individual filter values
def build_filter_string(filter_values):
    filter_string = ""
    index = 1
    for filter_value in filter_values:
        match_type = filter_value["matchtype"]
        value = parse_filter_value(filter_value["values"])
        filter_string += "%7B\"type\":\"" + match_type + \
            "\",\"values\":%5B%7B\"constant\":\"" + \
            value + "\"%7D,%7B%7D%5D%7D"
        if index < len(filter_values):
            filter_string += ","
        elif index == len(filter_values):
            filter_string += "%5D"
        index += 1
    return filter_string


# Set up flag for filters
filtered = False

# Set configuration
if (len(sys.argv) < 2):  # Will be 1 if no arguments, 2 if one argument
    print "Usage: python looker_embed_generator.py \
        <<embed url>> {\"filtername\": \"filtervalue\"}"
    sys.exit(1)

if (len(sys.argv) == 3):  # Will be 3 if passing in a json object of filters
    filtered = True
    filters = json.loads(sys.argv[2])

embed_url = '/embed/' + sys.argv[1]

# If the filtered flag is set, break out the filter names and values
# from the JSON string passed in via cms line, and add these to the
# embed_url string.
if filtered:
    embed_url += encode_embed_filters(filters)

lookerkey = os.getenv('LOOKERKEY')

if (lookerkey is None):  # confirm that the environment variable is set
    print "LOOKERKEY environment variable not set"
    sys.exit(1)

lookerurl = 'dev.analytics.gov.bc.ca'  # set to the URL where Looker is hosted


class Looker:
    def __init__(self, host, secret):
        self.secret = secret
        self.host = host


class User:
    def __init__(self, id=id, first_name=None, last_name=None,
                 permissions=[], models=[], group_ids=[],
                 external_group_id=None,
                 user_attributes={}, access_filters={}):
        self.external_user_id = json.dumps(id)
        self.first_name = json.dumps(first_name)
        self.last_name = json.dumps(last_name)
        self.permissions = json.dumps(permissions)
        self.models = json.dumps(models)
        self.access_filters = json.dumps(access_filters)
        self.user_attributes = json.dumps(user_attributes)
        self.group_ids = json.dumps(group_ids)
        self.external_group_id = json.dumps(external_group_id)


class URL:
    def __init__(self, looker, user, session_length,
                 embed_url, force_logout_login=False):
        self.looker = looker
        self.user = user
        self.path = '/login/embed/' + urllib.quote_plus(embed_url)
        self.session_length = json.dumps(session_length)
        self.force_logout_login = json.dumps(force_logout_login)

    # The current time as a UNIX timestamp.
    def set_time(self):
        self.time = json.dumps(int(time.time()))

    # Random string cannot be repeated within an hour. Prevents an
    # attacker from re-submitting a legitimate user's URL to gather
    # information they shouldn't have.
    def set_nonce(self):
        self.nonce = json.dumps(binascii.hexlify(os.urandom(16)))

    def sign(self):
        #  Do not change the order of these
        string_to_sign = ""
        string_to_sign = string_to_sign + self.looker.host + "\n"
        string_to_sign = string_to_sign + self.path + "\n"
        string_to_sign = string_to_sign + self.nonce + "\n"
        string_to_sign = string_to_sign + self.time + "\n"
        string_to_sign = string_to_sign + self.session_length + "\n"
        string_to_sign = string_to_sign + self.user.external_user_id + "\n"
        string_to_sign = string_to_sign + self.user.permissions + "\n"
        string_to_sign = string_to_sign + self.user.models + "\n"
        string_to_sign = string_to_sign + self.user.group_ids + "\n"
        string_to_sign = string_to_sign + self.user.external_group_id + "\n"
        string_to_sign = string_to_sign + self.user.user_attributes + "\n"
        string_to_sign = string_to_sign + self.user.access_filters

        signer = hmac.new(self.looker.secret,
                          string_to_sign.encode('utf8'), sha1)
        self.signature = base64.b64encode(signer.digest())

    def to_string(self):
        self.set_time()
        self.set_nonce()
        self.sign()

        params = {'nonce':               self.nonce,
                  'time':                self.time,
                  'session_length':      self.session_length,
                  'external_user_id':    self.user.external_user_id,
                  'permissions':         self.user.permissions,
                  'group_ids':           self.user.group_ids,
                  'models':              self.user.models,
                  'external_group_id':   self.user.external_group_id,
                  'user_attributes':     self.user.user_attributes,
                  'access_filters':      self.user.access_filters,
                  'signature':           self.signature,
                  'first_name':          self.user.first_name,
                  'last_name':           self.user.last_name,
                  'force_logout_login':  self.force_logout_login}

        query_string = '&'.join(["%s=%s" % (key,
                                            urllib.quote_plus(val)) for key,
                                val in params.iteritems()])

        return "%s%s?%s" % (self.looker.host, self.path, query_string)


def test():
    looker = Looker(lookerurl, lookerkey)

    user = User(50,
                first_name='Dashboard',
                last_name='User',
                permissions=['see_lookml_dashboards', 'access_data',
                             'see_user_dashboards', 'see_looks'],
                models=['all'],
                # group_ids=[1,54],
                external_group_id='external_group_id',
                # Add additional filters here. They must match a user
                # attribute that is listed in the LookML with an access_filter}
                user_attributes={"can_see_sensitive_data": "YES"}
                )

    # Set TTL for embed code. 60*15 = 15 minutes
    timeout = 60 * 15

    url = URL(looker, user, timeout, embed_url + ('&' if filtered else '?') +
              'embed_domain=http://localhost:8888', force_logout_login=True)

    return "https://" + url.to_string()


print test()
